doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1737,33 +1777,79 @@ Controlling the proof flow
    impossible. In most cases, :g:`P` or :g:`∼P` is one of the hypotheses of the
    local context.
 
-.. tacn:: contradiction
-   :name: contradiction
+.. tacn:: contradiction {? @one_term_with_bindings }
 
-   This tactic applies to any goal. The contradiction tactic attempts to
-   find in the current context (after all intros) a hypothesis that is
-   equivalent to an empty inductive type (e.g. :g:`False`), to the negation of
-   a singleton inductive type (e.g. :g:`True` or :g:`x=x`), or two contradictory
-   hypotheses :g:`P` and :g:`not P`.
+   Tries to prove the current goal by finding a contradiction.
 
-   .. exn:: No such assumption.
-      :undocumented:
+   If :n:`@one_term_with_bindings` is not provided (the most common use case),
+   the tactic first does an :tacn:`intros`.  The tactic then proves the goal if
+
+   - the updated context has a pair of hypotheses where one is the negation of
+     the other (e.g. :n:`P` and not :n:`~P`), or
+   - there is a hypothesis with an empty inductive hypothesis (e.g. :n:`False`), or
+   - there is a hypothesis that is the negation of a singleton inductive type
+     (e.g. :n:`True` or :n:`x=x`).
+
+   If :n:`@one_term_with_bindings` is provided, the type of
+   :n:`@one_term_with_bindings` must be a negation, such as :n:`~P`,
+   or :n:`False` (or another empty inductive type).
+   If the type is a negation and :n:`P` is a hypothesis in the context,
+   the goal is proven.  If the type is a negation and :n:`P` is not in
+   the context, the goal is replaced with :n:`P`.  If the type is :n:`False`,
+   the goal is proven.  Otherwise the tactic fails.  (In particular, note
+   that the converse doesn't work, e.g. if the type is :n:`P`, the tactic
+   won't find :n:`~P`.  Use :tacn:`contradict` if you want this behavior.)
```

Comment:
Actually, there are two "converse" cases here:

- if there is a hypothesis `P` and the user wants to replace the goal with `~P`, then `contradict` can be used instead
- if there are both hypotheses `H1 : P` and `H2 : ~P` in the context, the user can call `contradiction` without arguments or `contradiction H2`, but `contradiction H1` won't work (and `contradict` is irrelevant to this case).

Reply:
Perhaps the tactic should be extended to work for your H1/H2 case.  Should I file a PR?

Reply:
I think I've already expressed my opinion elsewhere that I think it's not worth changing. IMHO the most interesting use of `contradiction` is without arguments anyway.